### ISSUE TITLE
[Bug]: `useParameter` API is not working on initial load.

### ISSUE THREAD

**Issue by hwookim** [storybookjs/storybook#28592]
### Describe the bug

The `useParameter` API is not working on initial load.

I'm implementing and managing a `storybook-addon-cookie`.
My addon uses `useParameter` to read the cookie values entered by the user inside the story and display them in the panel.

However, since migrating to storybook version 8, when I access the parameter via `useParameter`, I get an empty value on initial load.
When I use `console.log` to `preview` and `Panel`, it appears to be getting an empty value until `preview` is called.

Is there any way to ensure that `preview.js` is called before `manager.js`?

```tsx
// Panel
export const PanelContent: React.FC = () => {
  const { path } = useStorybookState();
  const defaultCookie = useParameter<Cookie>(PARAM_KEY, {}); // TODO: Doesn't work on initial load

  const [story, setStory] = useState<string>();
  const [value, setValue] = useState<Cookie>();
  const [globals, updateGlobals] = useGlobals();

  const initCookie = () => {
    setStory(path);
    setValue(defaultCookie);
    clearCookies();
    setCookies(defaultCookie, encoding);
  };

  useEffect(() => {
    console.log('defaultCookie', defaultCookie);
    if (story !== path) {
      initCookie();
    }
  }, [defaultCookie, path, story]);
 //...
}
```
```ts
// preview
export const cookieDecorator = (
  storyFn: StoryFunction<Renderer>,
  { parameters }: DecoratorContext,
) => {
  const [flag, setFlag] = useState<boolean>(true);

  if (flag) {
    setFlag(false);
    if (!parameters) return storyFn();

    if (parameters[PARAM_PRESERVE_KEY] !== true) {
      clearCookies();
    }

    if (parameters[PARAM_KEY]) {
      console.log(parameters[PARAM_KEY]);
      setCookies(parameters.cookie, parameters.cookieEncoding);
    }
  }

  return storyFn();
};
```
![스크린샷 2024-07-14 오후 5 41 57](https://github.com/user-attachments/assets/45d5141c-a92f-4609-b355-3feb44d19be8)

### Reproduction link

https://github.com/hwookim/storybook-addon-cookie

### System

The package pretty much follows the settings in [addon-kit](https://github.com/storybookjs/addon-kit/blob/main/package.json).

**Comment by valentinpalkovic** [comment]
@ndelangen Any ideas?

**Comment by ndelangen** [comment]
> Is there any way to ensure that `preview.js` is called before `manager.js`?

Unfortunately, no.

This is due to the fact that parameters are dynamic, and stories are loaded async; the manager can only know about the parameters on a story, after it's been loaded.

**Comment by hwookim** [comment]
@ndelangen Is there any API to know if a parameter has been loaded?
It's hard to know by the presence or absence of a value, since the user can send an empty value.

**Comment by ndelangen** [comment]
No, there's currently no way to distinguish between those scenarios.

**Comment by Sidnioulz** [comment]
@valentinpalkovic this issue is what we were discussing. https://github.com/storybookjs/storybook/issues/22119 also relates to the same topic so this is an 'old' bug.

https://github.com/Sirrine-Jonathan/storybook-source-link/pull/21 can be used as a repro repository as it is visible: the first time you load a story in that SB instance, two addons disappear from the toolbar and the source-link addon fails to load its data.

You discussed a potential fix within `useParameter` that would ensure parameters get reloaded whenever a new story is prepared, therefore circumventing the race condition that @tmeasday discussed in his own issue. The starting point for the fix was:

```ts
useChannel(
    {
      [STORY_PREPARED]: () => {
        setStory(api.getCurrentStoryData());
      },
    },
    []
  );
```

**Comment by tmeasday** [comment]
Does it not eventually work when the parameters come over the wire? In my original issue it was just about a flash of "unparameterized" content, not the addon getting stuck without parameters.

**Comment by Sidnioulz** [comment]
> Does it not eventually work when the parameters come over the wire? In my original issue it was just about a flash of "unparameterized" content, not the addon getting stuck without parameters.

Yes and no. The add-ons don't get reloaded with all the parameters when the story loads. They're missing parameters until the second time the story is displayed (as, by then, the manager has loaded it).

**Comment by tmeasday** [comment]
Huh, I would have guessed `useParameters` would trigger a re-render when the parameters change (load), so there could be a bug there.

**Comment by Sidnioulz** [comment]
Just want to confirm that this is a regression (either SB 8 or SB 9). If I browse SB on https://github.com/Sidnioulz/storybook-source-link/tree/ee910f644134d43edc03963ab0d15b9d21133002 (which is version 7) and https://github.com/Sidnioulz/storybook-source-link/tree/main (version 9), I can see on v7 that the toolbar flickers the first time a story is rendered (but does not flicker in subsequent renders), demonstrating that it reloads once parameters become available. On the SB 9 branch, the toolbar no longer updates.

**Comment by valentinpalkovic** [comment]
Hi @Sidnioulz 

Could you please try out this canary and let me know whether it fixes the reactivity issue?
`0.0.0-pr-31579-sha-cf05dec5`
